### ISSUE TITLE
Map lookup parser failure

### ISSUE THREAD

**Issue by adamretter** [eXist-db/exist#2364]
The following query should return the second map, instead we get an error `XPST0003`. This does work in BaseX and Saxon, but not in eXist-db. Appears to be a parser error.

Tested on eXist-db 5.0.0-RC6

```xquery
let $ms := (
    map {
        "headers": (
            map {
                "name": "Content-Disposition",
                "value": 'filename="bob"'
            }
        )
    },
    map {
        "headers": (
            map {
                "name": "Content-Disposition",
                "value": 'filename="fred"'
            },
            map {
                "name": "Content-Type",
                "value": 'application/xml'
            }
        )
    }
)
return
    
$ms[not(empty(?headers[fn:matches(?value, 'filename="fred"')]))]
```

**Comment by duncdrum** [comment]
is this a separate problem from #2205, i can't really tell?

**Comment by adamretter** [comment]
@duncdrum yes, it's different.

**Comment by line-o** [comment]
It seems the parser already chokes on the implicit context element `.?headers` at least changes the error message.

The verbose equivalent of the code does run

```xquery
filter($ms, function ($m) {
    not(empty(
        filter($m?headers, function ($h) { 
            matches($h?value, 'filename="fred"')})))})
```

**Comment by adamretter** [comment]
@line-o Er.. I don't use `.` in the code above?!?

**Comment by line-o** [comment]
@adamretter yes, doing that just changes the error. That was all I wanted to add.

I tested two simplified queries:

```xquery
$ms[not(empty(?headers))]
``` 

> error found while executing expression: org.exist.xquery.XPathException: err:XPST0003 expecting closing parenthesis ')', found 'headers' [at line ...]

vs

```xquery
$ms[not(empty(.?headers))]
```

> Unexpectedly received 2 parameter(s) in call to function 'empty()'. Defined function signatures are: empty($items as item()*) xs:boolean

**Comment by line-o** [comment]
I am also wondering, if the same syntax for filtering node sets `//a[.]` also applies to sequences of maps, but I could not find anything. The xquery spec does refer to node-sets only, it seems.

@adamretter Maybe I am missing something, can you give me a pointer?

**Comment by line-o** [comment]
I can confirm it works in BaseX

**Comment by line-o** [comment]
exist 5.3.0-SNAPSHOT is affected

The originally reported issue seems to be limited to map lookups with implicit context passed to function calls inside of predicates.
```
(map {"f": 1}, map {"f": 2})[string(?f) eq '1']
```
whereas this works
```
(map {"f": 1}, map {"f": 2})[?f > 1]
```

**Comment by line-o** [comment]
With the changes to the XQuery parser included in eXist 5.3.0 you can write:
`$ms[?headers[?value ! matches(., 'filename="fred"')] => exists()]`

which is further proof that the unary lookup passed to a function call inside a predicate is the root cause of this issue.
